core/src/main/java/org/infinispan/factories/threads/EnhancedQueueExecutorFactory.java

```diff
@@ -0,0 +1,92 @@
+package org.infinispan.factories.threads;
+
+import static org.infinispan.commons.logging.Log.CONFIG;
+
+import java.time.Duration;
+import java.time.temporal.ChronoUnit;
+import java.util.concurrent.ThreadFactory;
+
+import org.infinispan.commons.executors.BlockingResource;
+import org.infinispan.commons.executors.BlockingThreadPoolExecutorFactory;
+import org.infinispan.commons.executors.ThreadPoolExecutorFactory;
+import org.infinispan.commons.util.concurrent.BlockingRejectedExecutionHandler;
+import org.jboss.threads.EnhancedQueueExecutor;
+import org.jboss.threads.management.ManageableThreadPoolExecutorService;
+
+public class EnhancedQueueExecutorFactory implements ThreadPoolExecutorFactory<ManageableThreadPoolExecutorService> {
+   private final int maxThreads;
+   private final int coreThreads;
+   private final int queueLength;
+   private final long keepAlive;
+
+   public EnhancedQueueExecutorFactory(int maxThreads, int coreThreads, int queueLength, long keepAlive, boolean nonBlocking) {
+      if (nonBlocking) {
+         throw new UnsupportedOperationException("EnhancedQueryExecutorFactory only supports blocking threads currently!");
+      }
+      this.maxThreads = maxThreads;
+      this.coreThreads = coreThreads;
+      this.queueLength = queueLength;
+      this.keepAlive = keepAlive;
+   }
+
+   public int maxThreads() {
+      return maxThreads;
+   }
+
+   public int coreThreads() {
+      return coreThreads;
+   }
+
+   public int queueLength() {
+      return queueLength;
+   }
+
+   public long keepAlive() {
+      return keepAlive;
+   }
+
+   public static EnhancedQueueExecutorFactory create(int maxThreads, int queueSize, boolean nonBlocking) {
+      int coreThreads = queueSize == 0 ? 1 : maxThreads;
+      return new EnhancedQueueExecutorFactory(maxThreads, coreThreads, queueSize,
+            BlockingThreadPoolExecutorFactory.DEFAULT_KEEP_ALIVE_MILLIS, nonBlocking);
+   }
+
+   @Override
+   public ManageableThreadPoolExecutorService createExecutor(ThreadFactory factory) {
+      if (!(factory instanceof BlockingResource)) {
+         throw new IllegalStateException("Executor factory configured to be blocking and received a thread" +
+               " factory that doesn't create blocking threads!");
+      }
+      EnhancedQueueExecutor.Builder builder = new EnhancedQueueExecutor.Builder();
+      builder.setThreadFactory(factory);
+      builder.setCorePoolSize(coreThreads);
+      builder.setMaximumPoolSize(maxThreads);
+      builder.setGrowthResistance(0.0f);
+      builder.setMaximumQueueSize(queueLength);
+      builder.setKeepAliveTime(Duration.of(keepAlive, ChronoUnit.MILLIS));
+
+      EnhancedQueueExecutor enhancedQueueExecutor = builder.build();
+      enhancedQueueExecutor.setHandoffExecutor(task ->
+         BlockingRejectedExecutionHandler.getInstance().rejectedExecution(task, enhancedQueueExecutor));
```

Comment:
Couldn't you make `BlockingRejectedExecutionHandler` implement `Executor`?

Reply:
I am not quite sure why we want that. We still need a reference to the original executor service to detect if it is shut down or not.

Reply:
I was thinking `BlockingRejectedExecutionHandler` doesn't need to implement `RejectedExecutionHandler` any more, so it might as well implement `Executor` and receive the `enhancedQueueExecutor` as a constructor parameter.

Reply:
Yeah, I personally am not a fan of that. Currently we can easily find all `RejectedExecutionHandler` implementations which seems much cleaner to me. And it very clearly documents what we should do with a task that is rejected from a blocking executor.

And the resulting class seems really ugly to me in that we create an Executor that takes an ExecutorService but can't submit to it, only check its running state. It seems like a dependency going the wrong direction, where as now it is a much clearer to me.

Reply:
Fair enough, it's complicated enough with all those `*ExecutorFactory` classes, no need for one more `Executor`.